File: leetcode/easy/69. Sqrt(x).ts
function mySqrt(x: number): number {
  if (x === 0 || x === 1) return x;

  // def binary_search(array) -> int:
  //     def condition(value) -> bool:
  //         pass

  //     left, right = 0, len(array)
  //     while left < right:
  //         mid = left + (right - left) // 2
  //         if condition(mid):
  //             right = mid
  //         else:
  //             left = mid + 1
  //     return left

  let start: number = 1;
  let end: number = x;
  while (start < end) {
    let mid: number = Math.floor((start + end) / 2);
    if (mid * mid > x) {
      end = mid;
    } else {
      start = mid + 1;
    }
  }
  return start - 1;

  let startIdx: number = 1;
  while (startIdx * startIdx <= x) {
    if (startIdx * startIdx === x) return startIdx;
    startIdx += 1;
    if (startIdx * startIdx > x) {
      startIdx -= 1;
      break;
    }
  }
  return startIdx;
}

File: leetcode/easy/69. Sqrt(x).rs
impl Solution {
    pub fn my_sqrt(x: i32) -> i32 {
        if x == 1 || x == 0{
            return x;
        }
        let mut start: i64 = 1;
        let mut end: i64 = (x/2+1) as i64;

        while start < end{
            let mid: i64 = start + (end-start)/2;
            if mid*mid > x as i64{
                end = mid;
                continue;
            }
            start = mid + 1;
        }

        return (start-1) as i32;
    }
}